lib/rpmsg/rpmsg_virtio.c

```diff
@@ -313,37 +313,55 @@ static int _rpmsg_virtio_get_buffer_size(struct rpmsg_virtio_device *rvdev)
 	return length;
 }
 
-static void rpmsg_virtio_hold_rx_buffer(struct rpmsg_device *rdev, void *rxbuf)
+static void rpmsg_virtio_hold_rx_buffer_nolock(struct rpmsg_device *rdev,
+					       struct rpmsg_hdr *rp_hdr)
 {
-	struct rpmsg_hdr *rp_hdr;
-
 	(void)rdev;
 
-	rp_hdr = RPMSG_LOCATE_HDR(rxbuf);
-
 	/* Set held status to keep buffer */
-	rp_hdr->reserved |= RPMSG_BUF_HELD;
+	rp_hdr->reserved += 1 << RPMSG_BUF_HELD_SHIFT;
 }
 
-static void rpmsg_virtio_release_rx_buffer(struct rpmsg_device *rdev,
-					   void *rxbuf)
+static void rpmsg_virtio_hold_rx_buffer(struct rpmsg_device *rdev, void *rxbuf)
+{
+	metal_mutex_acquire(&rdev->lock);
+	rpmsg_virtio_hold_rx_buffer_nolock(rdev, RPMSG_LOCATE_HDR(rxbuf));
+	metal_mutex_release(&rdev->lock);
+}
+
+static bool rpmsg_virtio_release_rx_buffer_nolock(struct rpmsg_device *rdev,
+						  struct rpmsg_hdr *rp_hdr)
 {
 	struct rpmsg_virtio_device *rvdev;
-	struct rpmsg_hdr *rp_hdr;
 	uint16_t idx;
 	uint32_t len;
 
 	rvdev = metal_container_of(rdev, struct rpmsg_virtio_device, rdev);
-	rp_hdr = RPMSG_LOCATE_HDR(rxbuf);
-	/* The reserved field contains buffer index */
-	idx = (uint16_t)(rp_hdr->reserved & ~RPMSG_BUF_HELD);
 
-	metal_mutex_acquire(&rdev->lock);
+	rp_hdr->reserved -= 1 << RPMSG_BUF_HELD_SHIFT;
```

Comment:
test first that ` rp_hdr->reserved` is not null

Reply:
Done